round-based/src/mpc/mod.rs

```diff
@@ -0,0 +1,329 @@
+
+    /// Returned by [`.send_many()`](Self::send_many)
+    type SendMany: SendMany<Exec = Self, Msg = Self::Msg, SendErr = Self::SendErr>;
+
+    /// Completes the round
+    ///
+    /// Waits until all messages in the round `R` are received, returns the received messages.
+    async fn complete<R>(
+        &mut self,
+        round: Self::Round<R>,
+    ) -> Result<R::Output, Self::CompleteRoundErr<R::Error>>
+    where
+        R: RoundInfo,
+        Self::Msg: RoundMsg<R::Msg>;
+
+    /// Sends a message
+    ///
+    /// This method awaits until the message is sent, which might be not the best method to use if you
+    /// need to send many messages at once. If it's the case, prefer using [`.send_many()`](Self::send_many).
+    async fn send(&mut self, msg: Outgoing<Self::Msg>) -> Result<(), Self::SendErr>;
+
+    /// Sends a p2p message to another party
+    ///
+    /// Note: when you send many messages at once (it's most likely the case when you send a p2p message), this method
+    /// is not efficient, prefer using [`.send_many()`](Self::send_many).
+    async fn send_p2p(
+        &mut self,
+        recipient: PartyIndex,
+        msg: Self::Msg,
+    ) -> Result<(), Self::SendErr> {
+        self.send(Outgoing::p2p(recipient, msg)).await
+    }
+
+    /// Sends a message that will be received by all parties
+    ///
+    /// Message will be broadcasted, but not reliably. If you need a reliable broadcast, use
+    /// [`MpcExecution::reliably_broadcast`] method.
+    async fn send_to_all(&mut self, msg: Self::Msg) -> Result<(), Self::SendErr> {
+        self.send(Outgoing::all_parties(msg)).await
+    }
+
+    /// Reliably broadcasts a message
+    ///
+    /// Message will be received by all participants of the protocol. Moreover, when recipient receives a
+    /// message, it will be assured (cryptographically or through other trust assumptions) that all honest
+    /// participants of the protocol received the same message.
+    ///
+    /// It's a responsibility of a message delivery layer to provide the reliable broadcast mechanism. If
+    /// it's not supported, this method returns an error. Note that not every MPC protocol requires the
+    /// reliable broadcast, so it's totally normal to have a message delivery implementation that does
+    /// not support it.
+    async fn reliably_broadcast(&mut self, msg: Self::Msg) -> Result<(), Self::SendErr> {
+        self.send(Outgoing::reliable_broadcast(msg)).await
+    }
+
+    /// Creates a buffer of outgoing messages so they can be sent all at once
+    ///
+    /// When you have many messages that you want to send at once, using [`.send()`](Self::send) is not efficient,
+    /// as it will accumulate message delivery cost. Use this method to optimize sending many messages at once.
+    fn send_many(self) -> Self::SendMany;
+
+    /// Yields execution
+    async fn yield_now(&self);
+}
+
+/// Buffer, optimized for sending many messages at once
+pub trait SendMany {
```

Comment:
I think this should explicitly mention the `MpcExecution` since those are closely tied. Like "this is a special temporary form of `MpcExecution` to send multiple unicast messages in a row"